File: src/spots.jl
using Random
using Statistics

export Spot, SpotDynamics, modulate

"""
Siderial rotation period at the equator
"""
const PROT_SUN = 24.5
const OMEGA_SUN = 2 * π / PROT_SUN / 86400

struct Spot{T <: Number}
    nday::Int # day index
    lat::T
    lon::T
    Φmax::T
    area_max::T
end

struct SpotDynamics{T <: Number}
    spots::Vector{<:Spot}
    duration::T
    inclination::T
    ω::T
    Δω::T
    equatorial_period::T
    τ_emergence::T
    τ_decay::T
end

"""
    SpotDynamics(spots::AbstractVector{Spot};
        duration = maximum([s.nday for s in spots]),
        alpha_med = 0.0001,
        inclination = asin(rand()),
        ω = 1.0,
        Δω = 0.2,
        τ_decay = 5.0,
        threshold = 0.1)

A container for the dynamics of starspots.

# Parameters
* spots - The list of `Spot`s evolved over time
* duration - The length of the evolution time
* alpha_med - An activation parameter for the magnetic flux
* inclination - Inclination of the star from our line of sight in radians
* ω - Rotational velocity of the star in solar units
* Δω - change in rotational velocity over the time in solar untis
* τ_decay - The decay timescale
* threshold - The threshold in magnetic flux for filtering starspots
"""
function SpotDynamics(spots::AbstractVector{Spot};
    duration = maximum([s.nday for s in spots]),
    alpha_med = 0.0001,
    inclination = asin(rand()),
    ω = 1.0,
    Δω = 0.2,
    τ_decay = 5.0,
    threshold = 0.1)
    
    ω *= OMEGA_SUN
    Δω *= OMEGA_SUN
    eq_per = 2π / ω / 86400
    τ_emergence = max(2.0, eq_per * τ_decay / 5)
    τ_decay *= eq_per


    filt(s) = (s.nday < duration) && (s.Φmax > threshold)
    spots = filter(filt, spots)
    Φmax = [s.Φmax for s in spots]
    area_max = alpha_med .* Φmax ./ median(Φmax)
    spots = [Spot(s.nday, s.lat, s.lon, s.Φmax, a) for (s, a) in zip(spots, area_max)]

    return SpotDynamics(spots, promote(duration, inclination, ω, Δω, eq_per, τ_emergence, τ_decay)...)
end

Base.broadcastable(r::SpotDynamics) = Ref(r)
Base.length(r::SpotDynamics) = length(r.spots)
Base.size(r::SpotDynamics) = size(r.spots)
Base.size(r::SpotDynamics, i) = size(r.spots, i)

function Base.show(io::IO, sd::SpotDynamics)
    out = """
    $(typeof(sd))
      nspots: $(length(sd))
      duration: $(sd.duration)
      inclination: $(sd.inclination)
      ω: $(sd.ω)
      Δω: $(sd.Δω)
      equatorial_period: $(sd.equatorial_period)
      τ_emergence: $(sd.τ_emergence)
      τ_decay: $(sd.τ_decay)"""
    print(io, out)
end
"""
    diffrot(ω₀, Δω, lat)

Default differental rotation function

Returns angular velocity as a function of latitude [0°, 90°]
"""
@inline diffrot(ω₀, Δω, lat) = ω₀ - Δω * sin(lat)^2

function modulate(spot::Spot, t,
    τ_emergence, τ_decay,
    ω, Δω, inclination, diffrot_func = diffrot)
    # Get spot area
    tt = t - spot.nday
    area = spot.area_max
    timescale = tt < 0 ? τ_emergence : τ_decay
    
    area *= exp(-(tt / timescale)^2 / 2)

    # rotation rate
    phase = diffrot(ω, Δω, spot.lat) * t * 86400 + spot.lon

    # foreshortening
    cos_beta = cos(inclination) * sin(spot.lat) + sin(inclination) * cos(spot.lat) * cos(phase)

    return -area * max(cos_beta, 0.0)
end

"""
    modulate(::SpotDynamics, time)

Modulate the flux due to starspots at the given timestep in days.
"""
function modulate(r::SpotDynamics, t)
    dFlux = sum(modulate.(r.spots, t, r.τ_emergence, r.τ_decay, r.ω, r.Δω, r.inclination))
    return dFlux
end

File: butterpy/spots.py
"""
spots.py
Contains the Spots class, which holds parameters for spots on a given star.
"""

import numpy as np
import matplotlib.pyplot as plt
from matplotlib.pyplot import subplots
from .constants import RAD2DEG, DAY2SEC, OMEGA_SUN


def diffrot(omega0, delta_omega, lat):
    """
    Default differential rotation function:
    Gives angular velocity as a function of latitude.
    """
    return omega0 - delta_omega * np.sin(lat) ** 2


class Spots(object):
    """
    Holds parameters for spots on a given star.

    PARAMETERS
    ----------
    spot_properties

    dur

    alpha_med

    incl

    omega

    delta_omega

    diffrot_func

    tau_evol

    threshold

    """

    def __init__(
        self,
        spot_properties,
        dur=None,
        alpha_med=0.0001,
        incl=None,
        omega=2.0,
        delta_omega=0.3,
        diffrot_func=diffrot,
        decay_timescale=5.0,
        threshold=0.1,
    ):

        ## global stellar parameters, same for all spots

        self.spot_properties = spot_properties
        if incl is None:
            self.inclination = np.arcsin(np.random.uniform())
        else:
            self.inclination = incl

        # rotation
        self.omega = omega * OMEGA_SUN
        self.delta_omega = delta_omega * OMEGA_SUN
        self.equatorial_period = 2 * np.pi / self.omega / DAY2SEC
        self.diffrot_func = diffrot_func

        # spot emergence and decay timescales
        self.emergence_timescale = max(
            2.0, self.equatorial_period * decay_timescale / 5
        )
        self.decay_timescale = self.equatorial_period * decay_timescale

        # convert active region properties
        time = spot_properties["nday"].values
        latitude = spot_properties["lat"].values
        longitude = spot_properties["lon"].values
        peak_magnetic_flux = spot_properties["bmax"]

        # keep only spots emerging within specified timespan and with peak B-field > threshold
        if dur is None:
            self.dur = time.max()
        else:
            self.dur = dur

        keep = (time < self.dur) & (peak_magnetic_flux > threshold)
        self.nspot = keep.sum()
        self.t0 = time[keep]
        self.latitude = latitude[keep]
        self.longitude = longitude[keep]
        # self.area_max = active_region_tilt[keep] * alpha_med / np.median(active_region_tilt[keep]) # median-divide to scale.
        # Aigrain et al. (2015) has this value as alpha_med*Bmax/mean(Bmax)
        self.area_max = (
            alpha_med * (peak_magnetic_flux / np.median(peak_magnetic_flux))[keep]
        )
        # alpha is spot contrast * spot area

    def calc(self, time):
        """Modulate flux using chunks for all spots"""
        N = len(time)

        if N > 365 * 48:
            nbins = 100
            n = int(N / nbins)
            dFlux = np.zeros(N)
            for i in range(nbins):
                dFlux[i * n : (i + 1) * n] = self.matrix_calc(time[i * n : (i + 1) * n])
        else:
            dFlux = self.matrix_calc(time)

        return dFlux

    def matrix_calc(self, time):
        """Modulate flux for all spots"""
        M = self.nspot
        N = len(time)

        # spot area
        tt = np.outer(np.ones(M), time) - np.outer(self.t0, np.ones(N))
        area = np.outer(self.area_max, np.ones(N))
        emerge = np.outer(self.emergence_timescale, np.ones(N))
        decay = np.outer(self.decay_timescale, np.ones(N))
        timescale = np.where(tt < 0, emerge, decay)
        area *= np.exp(-0.5 * (tt / timescale) ** 2)

        # rotation rate
        omega = self.diffrot_func(self.omega, self.delta_omega, self.latitude)
        # here 'phase' is the longitude as a function of time
        phase = np.outer(omega, time * DAY2SEC) + np.outer(self.longitude, np.ones(N))

        # foreshortening
        # if inclination is angle between pole and line of sight,
        # cos(beta) = cos(i)*sin(lat) + sin(i)*cos(lat)*cos(lon(t)).
        cos_beta = np.cos(self.inclination) * np.outer(
            np.sin(self.latitude), np.ones(N)
        ) + np.sin(self.inclination) * np.outer(
            np.cos(self.latitude), np.ones(N)
        ) * np.cos(
            phase
        )

        # Total modulation on stellar flux
        dFlux = -area * np.maximum(cos_beta, 0)

        # return area, omega, beta, dFlux
        return dFlux.sum(axis=0)


def butterfly(self):
    """Plot the butterfly diagram"""
    latitude = self.spot_properties["lat"]

    fig, ax = subplots()
    ax.scatter(
        self.spot_properties["nday"],
        latitude * RAD2DEG,
        s=10 * self.spot_properties["bmax"] / np.median(self.spot_properties["bmax"]),
        alpha=0.5,
        c="#996699",
        lw=0.5,
    )
    ax.set_xlim(0, self.spot_properties["nday"].max())
    ax.set_ylim(-90, 90)
    ax.set_yticks((-90, -45, 0, 45, 90))
    ax.set_xlabel("Time (days)")
    ax.set_ylabel("Latitude (deg)")
    fig.tight_layout()


def main():
    print('"main" functionality not implemented yet!')


if __name__ == "__main__":
    main()
